{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-23T19:12:58.554847Z",
     "start_time": "2021-11-23T19:12:54.606079Z"
    },
    "init_cell": true
   },
   "outputs": [],
   "source": [
    "### IMPORT LIBRARIES ###\n",
    "import os\n",
    "import math\n",
    "import magpy as mp\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "expt_path = \"/proj/magness/CDS010-014-015_combined\"\n",
    "contrast_palette = ['#00ff7f','#b22222','#b8b8b8','#f0e68c','#705000','#00bfff','#dda0dd','#00ffff','#ccccff',\n",
    "                    '#ff0000','#808000','#3cb371','#a97a67','#ff8c00','#ffd700','#02026e','#adff2f','#4169e1',\n",
    "                    '#ffa07a','#2f4f4f','#7f007f','#6e6e6e','#ff00ff']\n",
    "\n",
    "##Setting figure output parameters\n",
    "sc.set_figure_params(dpi_save = 300, dpi = 200, format = 'svg')\n",
    "sc.settings.figdir = expt_path+'/figures/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Figure 10A - \"Pan-Goblet Cell\" Heatmap for SI vs Colon GCs DEGs ###\n",
    "\n",
    "adata = mp.load(combined_path, \"clustered_annotated_adata_k25_lr0.92_v1.5.h5ad\") #Check if latest version\n",
    "adata = mp.scale_adata(adata)\n",
    "\n",
    "#goblet cells\n",
    "subset = adata[adata.obs.type=='goblet']\n",
    "\n",
    "gene_dict_gob = {\n",
    "    'Goblet_genes':['CLCA1',\t'SYTL2',\t'MUC2',\t'FCGBP',\t'REP15',\t'CREB3L1',\t'AGR2',\t'XBP1',\t'SPDEF',\t'GOLGB1',\t'ST6GALNAC1',\t'TM9SF3',\t'SMIM14',\t'KDELR2',\t'CDC42EP5',\t'TSPAN13',\t'HDLBP',\t'SPINK4',\t'ATOH1',\t'TFF3'],\n",
    "    'SI_gob_diff_genes':['ABCA4',\t'REG3A',\t'FAM3B',\t'CCL25',\t'TM4SF4',\t'REG4',\t'AQP3',\t'RAMP1',\t'GCLM',\t'SI',\t'UGT2B7',\t'MDK',\t'SLC16A7',\t'TM4SF5',\t'EIF4EBP1',\t'KRT20',\t'NQO1',\t'AKR1C3',\t'CYB5A',\t'SERPINB6'],\n",
    "    'C_gob_diff_genes':['MUC5B',\t'L1TD1',\t'NXPE4',\t'C10orf99',\t'VSIG2',\t'WFDC2',\t'BEST2',\t'CKB',\t'LDHB',\t'CA12',\t'MUC4',\t'NXPE1',\t'B3GNT6',\t'TRABD2A',\t'CEACAM5',\t'MUC1',\t'PLCB4',\t'AC103702.2',\t'IGFBP2',\t'CD24']\n",
    "}\n",
    "\n",
    "subset.obs.organ.cat.reorder_categories(['SI','Colon'], inplace=True)\n",
    "sc.pl.heatmap(subset, gene_dict_gob, groupby = 'organ' , swap_axes = True, show_gene_labels = True, cmap = mp.blues, standard_scale = 'var') #save = \"PanGobletCell_Heatmap\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Figure 10B - Mucins in SI vs. Colon by Region, All Lineages vs. GCs ###\n",
    "adata = mp.load(combined_path, \"clustered_annotated_adata_k25_lr0.92_v1.5.h5ad\") #Check if latest version\n",
    "\n",
    "#List of Mucin Genes in SI vs Colon GCs, reordered by trends\n",
    "subset = adata[adata.obs.type=='goblet']\n",
    "subset = mp.scale_adata(subset,groupby = 'region')\n",
    "gene_list = ['MUC2','MUC13','MUC17','MUC1','MUC4','MUC5B','MUC12','MUC3A','MUC20']\n",
    "\n",
    "# Figure 10C Dotplot of mucin gene expression across SI vs colon GOBLET cells\n",
    "print('mucins SI vs colon GCs')\n",
    "sc.pl.dotplot(subset, gene_list, groupby = 'region', swap_axes = True, layer='raw_scaled') #save = \"Mucins_SI_C_GCs_Region\" \n",
    "\n",
    "# Figure 10D otplot of mucin gene expression across SI AEs and Colon ACCs\n",
    "subset = adata[adata.obs.lineage.isin(['SI_earlyAE','SI_intermAE','SI_matureAE','C_earlyACC','C_lateACC'])]\n",
    "subset = mp.scale_adata(subset,groupby='region')\n",
    "print('mucins SI AEs vs colon ACCs')\n",
    "sc.pl.dotplot(subset, gene_list, groupby = 'region', swap_axes = True, layer='raw_scaled') #save = \"Mucins_AEs_ACCs_Region\" \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Figure 10E, 11A - GC subclustering\n",
    "\n",
    "## Goblet cell subclustering\n",
    "\n",
    "adata = mp.load(expt_path, 'clustered_annotated_adata_k25_lr0.92_v1.5.h5ad')\n",
    "\n",
    "mito_genes = adata.var_names.tolist()\n",
    "mito_genes[:] = [gene for gene in mito_genes if gene.startswith('MT-')]\n",
    "mito_genes.append('MALAT1')\n",
    "\n",
    "# switch below C tag for SI when looking at SI subclustering for supplemental figure 11\n",
    "gob_subset = adata[adata.obs.lineage.isin(['C_goblet','C_secretory_prog']),~adata.var_names.isin(mito_genes)]\n",
    "raw_subset = gob_subset.raw.to_adata()\n",
    "raw_subset = raw_subset[:,~adata.var_names.isin(mito_genes)]\n",
    "del gob_subset.raw\n",
    "gob_subset.raw = raw_subset\n",
    "\n",
    "#colon settings\n",
    "num_neighbors = 5\n",
    "num_pcs = 22\n",
    "leiden_res = 0.3\n",
    "hvg_num = 4000\n",
    "kwargs = {'max_iter_harmony':30}\n",
    "\n",
    "# ##SI settings\n",
    "# num_neighbors = 10\n",
    "# num_pcs = 19\n",
    "# leiden_res = 0.4\n",
    "# hvg_num = 2000\n",
    "# kwargs = {'max_iter_harmony':30}\n",
    "\n",
    "gob_subset = mp.pipeline.recluster(gob_subset, save=False, expt_path=expt_path, write_file='gob_subclustering_for_DE.h5ad', \n",
    "                                   use_raw = True, recalc_pca = True, reembed=True, regress_qc_params = False,\n",
    "                                   recalc_hvg = True, hvg_num = hvg_num, harmonize=True, annotate_cell_cycle=True, \n",
    "                                   regress_cell_cycle=False, num_pcs=num_pcs, num_neighbors=num_neighbors, \n",
    "                                   leiden_res=leiden_res, rescale_data = False, **kwargs)\n",
    "\n",
    "gob_subset2 = gob_subset[~gob_subset.obs.leiden.isin(['4','6'])]\n",
    "# gob_subset2 = gob_subset[~gob_subset.obs.leiden.isin([])]\n",
    "\n",
    "sc.tl.paga(gob_subset2, groups = 'leiden')\n",
    "sc.pl.paga(gob_subset2)\n",
    "sc.tl.umap(gob_subset2, init_pos = 'paga')\n",
    "\n",
    "print(gob_subset.obs.leiden.value_counts())\n",
    "\n",
    "print(\"Highly variable genes: %d\"%sum(gob_subset.var.highly_variable))\n",
    "\n",
    "sc.pl.umap(gob_subset2, color = ['leiden','MXD1','MUC2','CLCA1','FER1L6','APOB','MKI67','AGR2','FCGBP','SYTL2','SLFN5','ERN2','DEFA5','DEFA6','LYZ'], ncols = 2)\n",
    "\n",
    "sc.tl.rank_genes_groups(gob_subset2, 'leiden', method='wilcoxon')\n",
    "\n",
    "result = gob_subset2.uns['rank_genes_groups']\n",
    "groups = result['names'].dtype.names\n",
    "df = pd.DataFrame(\n",
    "    {group + '_' + key[:1]: result[key][group]\n",
    "    for group in groups for key in ['names', 'pvals']}).head(20)\n",
    "\n",
    "# mp.save(gob_subset2, combined_path, 'SI_gob_subclustering_for_DE.h5ad')\n",
    "\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Figure 10 E,F,G & 11A,C\n",
    "adata = mp.load(combined_path, 'C_gob_subclustering_for_DE.h5ad')\n",
    "\n",
    "Gob_dict = {'early_gob':['STMN1','HMGB2','TOP2A','TUBA1B','MKI67'],\n",
    "            'crGC':['PDIA4','PDIA6','PDIA3','AGR2','PDIA5'],\n",
    "            'classical_gob':['CLCA1','TFF3','FCGBP','ZG16','ATOH1','FFAR4','MUC2'],\n",
    "            'icGC':['RAB3B','RAB27A','FER1L6','MXD1','ERN2','SPDEF']\n",
    "           }\n",
    "\n",
    "mucins = ['MUC2','MUC13','MUC1','MUC4','MUC5B','MUC12','MUC3A','MUC17','MUC20']\n",
    "\n",
    "sc.tl.diffmap(adata)\n",
    "sc.pp.neighbors(adata, n_neighbors=10, use_rep='X_diffmap')\n",
    "\n",
    "adata.uns['leiden_colors'] = contrast_palette\n",
    "\n",
    "adata.obs['leiden_anno'] = adata.obs['leiden']\n",
    "adata.uns['leiden_anno_colors'] = contrast_palette\n",
    "\n",
    "###C Leiden Names\n",
    "C_leiden_order = ['icGC','crGC-1','crGC-2','crGC-3','earlyGC']\n",
    "adata.obs['leiden_anno'].cat.categories = ['crGC-2','crGC-3','crGC-1','earlyGC','icGC']\n",
    "adata.obs.leiden_anno.cat.reorder_categories(C_leiden_order, inplace=True)\n",
    "\n",
    "###SI leiden names\n",
    "# SI_leiden_order = ['mature GC','ERN2_high','REG1A_high','AGR2_high','earlyGC']\n",
    "# adata.obs['leiden_anno'].cat.categories = ['REG1A_high','AGR2_high','mature GC','ERN2_high','earlyGC']\n",
    "# adata.obs.leiden_anno.cat.reorder_categories(SI_leiden_order, inplace=True)\n",
    "\n",
    "sc.tl.paga(adata, groups = 'leiden_anno')\n",
    "sc.pl.paga(adata)\n",
    "sc.tl.draw_graph(adata, init_pos='umap')\n",
    "\n",
    "root_cells = np.where(adata.obs['leiden_anno'] == 'earlyGC')[0]\n",
    "\n",
    "root_cell = np.random.choice(root_cells)\n",
    "\n",
    "adata.uns['iroot'] = root_cell\n",
    "\n",
    "sc.tl.dpt(adata)\n",
    "\n",
    "##Fig 10E, 11A\n",
    "sc.pl.umap(adata, color = ['leiden_anno'], save = '_supp_SI_gob_subset_leiden')\n",
    "##Fig 10F,G & 11A\n",
    "# sc.pl.umap(adata, color = ['dpt_pseudotime','MUC2'], save = '_supp-SI_gob_subset_dpt_MUC2')\n",
    "\n",
    "adata = mp.scale_adata(adata, groupby = 'leiden_anno')\n",
    "\n",
    "\n",
    "##Figure 10H & 11B or 10J,11C if you switch out Gob_dict for mucins list above\n",
    "\n",
    "# sc.pl.dotplot(adata, Gob_dict, groupby = 'leiden_anno', swap_axes = False, layer = 'raw_scaled', save = 'fig5_C_gob_subset_maturity')\n"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "kernelspec": {
   "display_name": "RNAseq Python",
   "language": "python",
   "name": "rnaseq"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
