{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-08-05T16:38:41.428478Z",
     "start_time": "2021-08-05T16:38:41.348567Z"
    },
    "init_cell": true
   },
   "outputs": [],
   "source": [
    "# Import libraries\n",
    "import os\n",
    "import math\n",
    "import magpy as mp\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "import diffxpy.api as de\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "combined_path = \"/proj/magness/CDS010-014-015_combined\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "### CHANGELOG ###\n",
    "\n",
    "# v1.2\n",
    "# -> Added .obs['organ'] annotation to specify SI v Colon\n",
    "# -> Removed .obs['type']== '15-?' cluster, comprised of high mitochondrial counts\n",
    "# -> Changed .obs['type']=='21-?' cluster to M_cells\n",
    "# -> Changed .uns['leiden_colors'] and .uns['type_colors'] to new colors and made consistent\n",
    "# -> Randomly shuffled cell indexes to make better overlays\n",
    "# -> Changed .obs['donor'] to 'Donor 1', 'Donor 2', and 'Donor 3'\n",
    "# -> Added .obs['cds'] containing 'CDS010', 'CDS014', and 'CDS015'\n",
    "\n",
    "# v1.3\n",
    "# -> Updated lineage to separate EEC and goblet_prog to SI_EEC/C_EEC and SI_goblet_proj/C_goblet_proj\n",
    "# -> Reordered lineage categories for easier plotting\n",
    "# -> Reordered Donor for numerical order\n",
    "# -> Reordered region by proximal to distal\n",
    "# -> Added in palettes for select .obs annotations\n",
    "# -> Added layer for raw annotations\n",
    "# -> Update colors for each lineage, organ, and donor and fig 1c\n",
    "\n",
    "# v1.4\n",
    "# -> Changed names in .obs['leiden']: Paneth/PLC -> SI_Best4/C_Best4 | goblet -> SI_secretory\n",
    "# -> Changed names in .obs['lineage']: Paneth/PLC -> SI_Best4/C_Best4 | split SI_secretory -> paneth/goblet\n",
    "# -> Changed .obs['type'] labels to Best4 and paneth\n",
    "# -> Added annotation for Parikh et al 2019 crypt-axis score\n",
    "# -> Added additional color for lineage=='paneth'\n",
    "# -> Added annotation of ['XXXXXXXX'] for donor+region adata.obs['donor_region'] = adata.obs.donor.str.cat(adata.obs.region)\n",
    "\n",
    "# v1.5\n",
    "# -> Added pseudotime info\n",
    "# -> Added raw_scaled data to object\n",
    "# -> Removed SI_paneth lineage\n",
    "# -> Cluster name updates in .obs['lineage']: SI-6? > SI_AE2, SI_ & C_goblet_prog > SI_sec_prog & C_sec_prog, C_earlyCC > C_earlyACC, C_latecc > C_lateACC\n",
    "\n",
    "# v1.6\n",
    "# -> update file name to be k10 NOT k25\n",
    "# -> add types for Secretory prog, AE2?\n",
    "# -> consolidating and reorganizing annotations for several var annotations that are duplicated across donors\n",
    "# -> concatenate gene_ids, feature_types, genome, mt\n",
    "# -> remove ribo, hb, all of .varm and .uns for cell type specific annotations\n",
    "# -> add donor annotation to n_cells, n_cells_by_counts, mean_counts, pct_dropout, total counts\n",
    "# -> move .raw layer to ln_rpdm-normalized and raw_normalized to raw_rpdm-normalized\n",
    "# -> add combined and preprocessed dataset to .raw layer\n",
    "\n",
    "# v1.7\n",
    "# -> deleted old aspects from previous version [adata.raw, adata.layers['raw','raw_rpdm-normalized', and 'ln_rpdm-normalized']\n",
    "# -> consolidated .var genes that shared ENSG ids based on the most recent annotation. these genes were summed in .raw and the raw_rpdm-normalized and ln_rpdm-normalized layers were updated accordingly\n",
    "# -> the values in adata.X which is the default layer for dimensional reduction, clustering, UMAP, etc were not adjusted. Only duplicate values were removed to correspond to the other updates from above  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "### Update of main data object to v1.2 ###\n",
    "### Changes for v1.2 ###\n",
    "\n",
    "adata = mp.load(combined_path, 'clustered_adata_k25_lr0.92.h5ad')\n",
    "print(adata.uns['neighbors'])\n",
    "# contrast_palette = ['#ff0000','#2f4f4f','#808000','#b22222','#3cb371','#7f007f','#d3d3d3','#ff8c00',\n",
    "#                     '#ffd700','#0000cd','#00ff7f','#4169e1','#ffa07a','#00bfff','#adff2f','#ff00ff',\n",
    "#                     '#f0e68c','#dda0dd','#ff1493','#00ffff']\n",
    "\n",
    "# adata.uns['type_colors'] = contrast_palette\n",
    "# adata.uns['leiden_colors'] = contrast_palette\n",
    "\n",
    "# print(adata)\n",
    "\n",
    "# print(adata.obs['donor'].value_counts())\n",
    "\n",
    "# adata = adata[~adata.obs['lineage'].isin(['15-?'])]\n",
    "\n",
    "# lineage_map = {'0':'SI_matureAE',\n",
    "#               '1':'SI_intermAE',\n",
    "#               '2':'C_earlyCC',\n",
    "#               '3':'SI_earlyAE',\n",
    "#               '4':'C_goblet',\n",
    "#               '5':'SI_ISC',\n",
    "#               '6':'SI_6-?',\n",
    "#               '7':'C_lateCC',\n",
    "#               '8':'C_tuft',\n",
    "#               '9':'C_TA',\n",
    "#               '10':'C_plc',\n",
    "#               '11':'SI_goblet',\n",
    "#               '12':'SI_TA',\n",
    "#               '13':'C_ISC',\n",
    "#               '14':'SI_paneth',\n",
    "#               '16':'SI_tuft',\n",
    "#               '17':'EEC',\n",
    "#               '18':'goblet_prog',\n",
    "#               '19':'SI_TA2',\n",
    "#               '21':'SI_FAE',\n",
    "#               }\n",
    "# adata.obs['lineage'] = adata.obs['leiden'].map(lineage_map)      \n",
    "\n",
    "# print(adata.obs['lineage'].value_counts())\n",
    "\n",
    "# type_map = {\n",
    "#     'SI_matureAE':'absorptive',\n",
    "#     'SI_intermAE':'absorptive',\n",
    "#     'C_earlyCC':'absorptive',\n",
    "#     'SI_earlyAE':'absorptive',\n",
    "#     'C_goblet':'goblet',\n",
    "#     'SI_ISC':'ISC',\n",
    "#     'SI_6-?':'SI_6-?',\n",
    "#     'C_lateCC':'absorptive',\n",
    "#     'C_tuft':'tuft',\n",
    "#     'C_TA':'TA',\n",
    "#     'C_plc':'paneth+plc',\n",
    "#     'SI_goblet':'goblet',\n",
    "#     'SI_TA':'TA',\n",
    "#     'C_ISC':'ISC',\n",
    "#     'SI_paneth':'paneth+plc',\n",
    "#     'SI_tuft':'tuft',\n",
    "#     'EEC':'EEC',\n",
    "#     'goblet_prog':'goblet',\n",
    "#     'SI_TA2':'TA',\n",
    "#     'SI_FAE':'FAE',\n",
    "#               }\n",
    "\n",
    "# adata.obs['type'] = adata.obs['lineage'].map(type_map) \n",
    "\n",
    "# print(adata.obs['type'].value_counts())\n",
    "\n",
    "# sc.pl.umap(adata, color = ['lineage','type'], palette = contrast_palette, ncols = 1)\n",
    "\n",
    "# mask1 = adata.obs['region'].isin(['Duo','Jej','Ile'])\n",
    "# mask2 = ~adata.obs['lineage'].str.startswith('C_')\n",
    "# is_SI = mask1 & mask2\n",
    "\n",
    "# mask3 = adata.obs['region'].isin(['AC','TC','DC'])\n",
    "# mask4 = ~adata.obs['lineage'].str.startswith('SI_')\n",
    "# is_colon = mask3 & mask4\n",
    "\n",
    "# adata.obs['organ'] = is_SI.map({True:'SI',False:'Colon'})\n",
    "\n",
    "# adata = adata[((is_SI) | (is_colon)),:]\n",
    "\n",
    "# sc.pl.umap(adata, color = 'organ')\n",
    "\n",
    "# print(adata.obs['organ'].value_counts())\n",
    "\n",
    "# adata.obs['cds'] = adata.obs['donor']\n",
    "\n",
    "# donor_map = {'CDS010':'Donor 2',\n",
    "#               'CDS014':'Donor 1',\n",
    "#               'CDS015':'Donor 3'\n",
    "#               }\n",
    "\n",
    "# adata.obs['donor'] = adata.obs['donor'].map(donor_map) \n",
    "\n",
    "# print(adata.obs['cds'].value_counts())\n",
    "# print(adata.obs['donor'].value_counts())\n",
    "\n",
    "# sc.pl.umap(adata, color = 'donor')\n",
    "\n",
    "# new_index = np.random.permutation(adata.obs_names)\n",
    "# print(new_index)\n",
    "# adata = adata[new_index,:]\n",
    "\n",
    "# sc.pl.umap(adata, color = 'donor')\n",
    "\n",
    "# mp.save(adata, combined_path, 'clustered_annotated_adata_k25_lr0.92_v1.2.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-04T16:16:15.133008Z",
     "start_time": "2021-06-04T16:16:09.723187Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "### Update of main data object to v1.3 ###\n",
    "\n",
    "expt_path = \"/proj/magness/CDS010-014-015_combined\"\n",
    "adata = mp.load(expt_path, \"clustered_annotated_adata_k25_lr0.92_v1.2.h5ad\")\n",
    "\n",
    "\n",
    "### CHANGE NAMES OF LEIDEN CLUSTERS ###\n",
    "lineage_map = {'0':'SI_matureAE',\n",
    "              '1':'SI_intermAE',\n",
    "              '2':'C_earlyCC',\n",
    "              '3':'SI_earlyAE',\n",
    "              '4':'C_goblet',\n",
    "              '5':'SI_ISC',\n",
    "              '6':'SI_6-?',\n",
    "              '7':'C_lateCC',\n",
    "              '8':'C_tuft',\n",
    "              '9':'C_TA',\n",
    "              '10':'C_plc',\n",
    "              '11':'SI_goblet',\n",
    "              '12':'SI_TA',\n",
    "              '13':'C_ISC',\n",
    "              '14':'SI_paneth',\n",
    "              '16':'SI_tuft',\n",
    "              '17':'EEC',\n",
    "              '18':'goblet_prog',\n",
    "              '19':'SI_TA2',\n",
    "              '21':'SI_FAE',\n",
    "              }\n",
    "\n",
    "adata.obs['leiden'] = adata.obs['leiden'].map(lineage_map)\n",
    "\n",
    "\n",
    "### SPLIT .OBS.LINEAGE EEC AND GOLBET_PROG INTO SI AND COLON ###\n",
    "colon_GPs = (adata.obs['lineage']=='goblet_prog') & (adata.obs['organ']=='Colon' )\n",
    "SI_GPs = (adata.obs['lineage']=='goblet_prog') & (adata.obs['organ']=='SI' )\n",
    "\n",
    "adata.obs['lineage'].cat.add_categories(new_categories=['C_goblet_prog','SI_goblet_prog'],inplace=True)\n",
    "adata.obs['lineage'] = adata.obs['lineage'].mask(colon_GPs,other='C_goblet_prog')\n",
    "adata.obs['lineage'] = adata.obs['lineage'].mask(SI_GPs,other='SI_goblet_prog')\n",
    "adata.obs['lineage'].cat.remove_categories(removals='goblet_prog',inplace=True)\n",
    "\n",
    "colon_EECs = (adata.obs['lineage']=='EEC') & (adata.obs['organ']=='Colon' )\n",
    "SI_EECs = (adata.obs['lineage']=='EEC') & (adata.obs['organ']=='SI' )\n",
    "\n",
    "adata.obs['lineage'].cat.add_categories(new_categories=['C_EEC','SI_EEC'],inplace=True)\n",
    "adata.obs['lineage'] = adata.obs['lineage'].mask(colon_EECs,other='C_EEC')\n",
    "adata.obs['lineage'] = adata.obs['lineage'].mask(SI_EECs,other='SI_EEC')\n",
    "adata.obs['lineage'].cat.remove_categories(removals='EEC',inplace=True)\n",
    "\n",
    "\n",
    "### REORDER LINEAGES AND DONOR LIST ###\n",
    "lineage_order = ['SI_ISC','C_ISC','SI_TA','SI_TA2','C_TA',\n",
    "                 'SI_earlyAE','C_earlyCC','SI_intermAE','SI_matureAE','SI_6-?','C_lateCC', \n",
    "                 'SI_paneth','C_plc','SI_tuft','C_tuft',\n",
    "                 'SI_goblet_prog','C_goblet_prog','SI_goblet','C_goblet',\n",
    "                 'SI_EEC','C_EEC','SI_FAE']\n",
    "\n",
    "adata.obs.lineage.cat.reorder_categories(lineage_order, inplace=True)\n",
    "\n",
    "donor_order = ['Donor 1','Donor 2','Donor 3']\n",
    "\n",
    "adata.obs.donor.cat.reorder_categories(donor_order, inplace=True)\n",
    "\n",
    "region_order = ['Duo','Jej','Ile','AC','TC','DC']\n",
    "\n",
    "adata.obs.region.cat.reorder_categories(region_order,inplace=True)\n",
    "\n",
    "### ADD RAW_NORMALIZED LAYER TO ADATA OBJECT ###\n",
    "adata2 = mp.load(expt_path,2)\n",
    "\n",
    "adata2 = adata2[adata.obs_names, adata.var_names]\n",
    "\n",
    "adata.layers['raw'] = adata2.X\n",
    "\n",
    "sc.pp.normalize_total(adata2)\n",
    "adata.layers['raw_normalized'] = adata2.X\n",
    "\n",
    "### SAVE ###\n",
    "mp.save(adata,expt_path,\"clustered_annotated_adata_k25_lr0.92_v1.3.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-23T15:48:29.708816Z",
     "start_time": "2021-07-23T15:48:07.463178Z"
    },
    "code_folding": [
     6,
     62
    ],
    "tags": []
   },
   "outputs": [],
   "source": [
    "### Update of main data object to v1.4 ###\n",
    "\n",
    "expt_path = \"/proj/magness/CDS010-014-015_combined\"\n",
    "adata = mp.load(expt_path, \"clustered_annotated_adata_k25_lr0.92_v1.3.h5ad\")\n",
    "\n",
    "### Change names in .obs['leiden']: Paneth/PLC -> SI_Best4/C_Best4 | SI_goblet -> SI_secretory | goblet_prog > secretory_prog\n",
    "leiden_map = {'SI_paneth':'SI_BEST4',\n",
    "              'SI_goblet':'SI_secretory',\n",
    "              'C_plc':'C_BEST4',\n",
    "              'goblet_prog':'secretory_prog',\n",
    "              'SI_matureAE':'SI_matureAE',\n",
    "              'SI_intermAE':'SI_intermAE',\n",
    "              'C_earlyCC':'C_earlyCC',\n",
    "              'SI_earlyAE':'SI_earlyAE',\n",
    "              'C_goblet':'C_goblet',\n",
    "              'SI_ISC':'SI_ISC',\n",
    "              'SI_6-?':'SI_6-?',\n",
    "              'C_lateCC':'C_lateCC',\n",
    "              'C_tuft':'C_tuft',\n",
    "              'C_TA':'C_TA',\n",
    "              'SI_TA':'SI_TA',\n",
    "              'C_ISC':'C_ISC',\n",
    "              'SI_tuft':'SI_tuft',\n",
    "              'EEC':'EEC',\n",
    "              'SI_TA2':'SI_TA2',\n",
    "              'SI_FAE':'SI_FAE',\n",
    "              }\n",
    "\n",
    "adata.obs['leiden'] = adata.obs['leiden'].map(leiden_map)\n",
    "\n",
    "### Change names in .obs['lineage']: Paneth/PLC -> SI_Best4/C_Best4 | split SI_secretory -> paneth/goblet\n",
    "# Subclustering and recalling paneth cells out of goblet cells\n",
    "\n",
    "mp.settings.leiden_resolution = 0.15\n",
    "mp.settings.num_neighbors = 15\n",
    "mp.settings.num_pcs = 40\n",
    "kwargs = {'max_iter_harmony':20}\n",
    "\n",
    "subset = adata[adata.obs['leiden']=='SI_secretory'].copy()\n",
    "subset = mp.pipeline.recluster(expt_path=combined_path, data=subset,\n",
    "                               write_file = 'goblet_subscluster_lr0.2_nn40_pcs40.h5ad',\n",
    "                               neighbors_key=None, harmonize=True, calc_leiden = True, reprocess = False,\n",
    "                               calc_hvg = True, annotate_cell_cycle = False, regress_cell_cycle = False,\n",
    "                               reembed = False, save = False, recalc_pca = False, **kwargs)\n",
    "\n",
    "# sc.pl.umap(subset, color = ['leiden','lineage','DEFA5','DEFA6','MUC2','LYZ'], ncols = 1)\n",
    "\n",
    "goblet_subset = subset[subset.obs['leiden']=='0']\n",
    "paneth_subset = subset[subset.obs['leiden']=='1']\n",
    "\n",
    "adata.obs.lineage.cat.add_categories([\"SI_BEST4\",\"C_BEST4\"], inplace = True)                   \n",
    "\n",
    "adata.obs['lineage'].mask(adata.obs['lineage']=='SI_paneth', 'SI_BEST4', inplace = True)\n",
    "adata.obs['lineage'].mask(adata.obs['lineage']=='C_plc', 'C_BEST4', inplace = True)\n",
    "adata.obs['lineage'].mask(adata.obs_names.isin(goblet_subset.obs_names.tolist()), 'SI_goblet', inplace = True)\n",
    "adata.obs['lineage'].mask(adata.obs_names.isin(paneth_subset.obs_names.tolist()), 'SI_paneth', inplace = True)\n",
    "\n",
    "adata.obs.lineage.cat.remove_categories(['C_plc'], inplace = True) \n",
    "\n",
    "### Change .obs['type'] labels to BEST4 and paneth and rename goblet prog to secretory prog\n",
    "# adata.obs.lineage.cat.rename_categories({'SI_goblet_prog':'SI_secretory_prog','C_goblet_prog':'C_secretory_prog','paneth+plc':'paneth'},inplace = True)     \n",
    "\n",
    "type_map = {\n",
    "    'SI_matureAE':'absorptive',\n",
    "    'SI_intermAE':'absorptive',\n",
    "    'C_earlyCC':'absorptive',\n",
    "    'SI_earlyAE':'absorptive',\n",
    "    'C_goblet':'goblet',\n",
    "    'SI_ISC':'ISC',\n",
    "    'SI_6-?':'SI_6-?',\n",
    "    'C_lateCC':'absorptive',\n",
    "    'C_tuft':'tuft',\n",
    "    'C_TA':'TA',\n",
    "    'SI_BEST4':'BEST4',\n",
    "    'C_BEST4':'BEST4',\n",
    "    'SI_goblet':'goblet',\n",
    "    'SI_TA':'TA',\n",
    "    'C_ISC':'ISC',\n",
    "    'SI_paneth':'paneth',\n",
    "    'SI_tuft':'tuft',\n",
    "    'SI_EEC':'EEC',\n",
    "    'C_EEC':'EEC',\n",
    "    'SI_secretory_prog':'secretory_prog',\n",
    "    'C_secretory_prog':'secretory_prog',\n",
    "    'SI_TA2':'TA',\n",
    "    'SI_FAE':'FAE'\n",
    "}\n",
    "\n",
    "adata.obs['type'] = adata.obs['lineage'].map(type_map)\n",
    "\n",
    "# sc.pl.umap(adata, color = 'type')\n",
    "# print(adata.obs.type.unique())\n",
    "\n",
    "### Added donor_region annotation ###\n",
    "\n",
    "adata.obs['donor_region'] = adata.obs.donor.str.cat(adata.obs.region,sep=\"_\")\n",
    "\n",
    "### ANNOTATE DATA WITH MEAN EXPRESSION, PCTEXP, and NCELLS ###\n",
    "import gc\n",
    "\n",
    "raw_adata = adata.copy()\n",
    "raw_adata.X = raw_adata.layers['raw_normalized'].todense()\n",
    "\n",
    "for lineage in adata.obs.lineage.unique():\n",
    "#     print()\n",
    "#     print(lineage)\n",
    "    gc.collect()\n",
    "    lineage_subset = raw_adata[raw_adata.obs.lineage == lineage]\n",
    "    adata.varm[f'{lineage}_means'] = np.zeros((adata.n_vars, 4, 3)) #gene, donor, region\n",
    "    adata.varm[f'{lineage}_pctexp'] = np.zeros((adata.n_vars, 4, 3)) #gene, donor, region\n",
    "    adata.uns[f'{lineage}_ncells'] = np.zeros((4, 3)) #donor, region\n",
    "    \n",
    "    if 'SI_' in lineage: regions = ['Duo','Jej','Ile']\n",
    "    else: regions = ['AC','TC','DC']\n",
    "    \n",
    "    for i, donor in enumerate(['combined','Donor 1','Donor 2','Donor 3']):\n",
    "#         print(donor)\n",
    "        if i == 0: donor_subset = lineage_subset\n",
    "        else: donor_subset = lineage_subset[lineage_subset.obs['donor']==donor]\n",
    "            \n",
    "        for j, region in enumerate(regions):\n",
    "#             print(region)\n",
    "            region_subset = donor_subset[donor_subset.obs['region']==region].copy()\n",
    "            \n",
    "            adata.varm[f'{lineage}_means'][:,i,j] = region_subset.X.mean(axis=0)\n",
    "            adata.varm[f'{lineage}_pctexp'][:,i,j] = (region_subset.X > 0).sum(axis=0) / region_subset.n_obs\n",
    "            adata.uns[f'{lineage}_ncells'][i,j] = region_subset.n_obs\n",
    "\n",
    "\n",
    "### SAVE NEW DATA VERSION ###            \n",
    "mp.save(adata,expt_path,\"clustered_annotated_adata_k25_lr0.92_v1.4.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Update of main data object to v1.5\n",
    "combined_path = \"/proj/magness/CDS010-014-015_combined\"\n",
    "adata = mp.load(combined_path, \"clustered_annotated_adata_k25_lr0.92_v1.42.h5ad\")\n",
    "\n",
    "### -> Cluster name updates in .obs['lineage']: \n",
    "## SI-6? > SI_AE2\n",
    "## SI_ & C_goblet_prog > SI_sec_prog & C_sec_prog\n",
    "## C_earlyCC > C_earlyACC\n",
    "## C_latecc > C_lateACC\n",
    "\n",
    "#print(adata.obs.lineage.value_counts())\n",
    "\n",
    "lineage_map = {'SI_paneth':'SI_paneth',\n",
    "               'SI_goblet':'SI_goblet',\n",
    "               'C_BEST4':'C_BEST4',\n",
    "               'SI_BEST4':'SI_BEST4',\n",
    "               'SI_goblet_prog':'SI_secretory_prog',\n",
    "               'C_goblet_prog':'C_secretory_prog',\n",
    "               'SI_matureAE':'SI_matureAE',\n",
    "               'SI_intermAE':'SI_intermAE',\n",
    "               'C_earlyCC':'C_earlyACC',\n",
    "               'SI_earlyAE':'SI_earlyAE',\n",
    "               'C_goblet':'C_goblet',\n",
    "               'SI_ISC':'SI_ISC',\n",
    "               'SI_6-?':'SI_AE2',\n",
    "               'C_lateCC':'C_lateACC',\n",
    "               'C_tuft':'C_tuft',\n",
    "               'C_TA':'C_TA',\n",
    "               'SI_TA':'SI_TA',\n",
    "               'C_ISC':'C_ISC',\n",
    "               'SI_tuft':'SI_tuft',\n",
    "               'SI_EEC':'SI_EEC',\n",
    "               'C_EEC':'C_EEC',\n",
    "               'SI_TA2':'SI_TA2',\n",
    "               'SI_FAE':'SI_FAE',\n",
    "              }\n",
    "\n",
    "adata.obs.lineage = adata.obs.lineage.map(lineage_map)\n",
    "\n",
    "#adata.obs.lineage.cat.remove_categories(['SI_goblet_prog','C_goblet_prog','C_earlyCC','C_lateCC','SI_6-?'], inplace = True)\n",
    "\n",
    "print(adata.obs.lineage.value_counts())\n",
    "\n",
    "sc.pl.umap(adata, color = 'lineage')\n",
    "# sc.pl.dotplot(adata, 'LGR5', groupby = 'lineage')\n",
    "\n",
    "mp.save(adata, combined_path, 'clustered_annotated_adata_k25_lr0.92_v1.5.h5ad')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### Update of main data object to v1.6\n",
    "adata = mp.load(combined_path, \"clustered_annotated_adata_k25_lr0.92_v1.5.h5ad\")\n",
    "print(adata)\n",
    "\n",
    "### -> renaming varm for\n",
    "## SI-6? > SI_AE2\n",
    "## SI_ & C_goblet_prog > SI_sec_prog & C_sec_prog\n",
    "## C_earlyCC > C_earlyACC\n",
    "## C_latecc > C_lateACC\n",
    "### consolidating and reorganizing annotations for several var annotations that are duplicated across donors\n",
    "#concatenate gene_ids, feature_types, genome, mt\n",
    "#remove ribo, hb\n",
    "#add donor annotation to n_cells, n_cells_by_counts, mean_counts, pct_dropout, total counts\n",
    "var_list = ['gene_ids', 'feature_types', 'genome','mt', 'n_cells','ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts','C_BEST4_means',\n",
    "            'C_BEST4_pctexp','C_EEC_means','C_EEC_pctexp','C_ISC_means','C_ISC_pctexp','C_TA_means','C_TA_pctexp','C_earlyCC_means','C_earlyCC_pctexp',\n",
    "            'C_goblet_means','C_goblet_pctexp','C_lateCC_means','C_lateCC_pctexp','C_tuft_means','C_tuft_pctexp','SI_BEST4_means','SI_BEST4_pctexp',\n",
    "            'SI_EEC_means','SI_EEC_pctexp','SI_FAE_means','SI_FAE_pctexp','SI_ISC_means','SI_ISC_pctexp','SI_TA2_means','SI_TA2_pctexp','SI_TA_means',\n",
    "            'SI_TA_pctexp','SI_earlyAE_means','SI_earlyAE_pctexp','SI_goblet_means','SI_goblet_pctexp','SI_intermAE_means','SI_intermAE_pctexp',\n",
    "            'SI_matureAE_means','SI_matureAE_pctexp','SI_paneth_means','SI_paneth_pctexp','SI_tuft_means','SI_tuft_pctexp','SI_6-?_means','SI_6-?_pctexp',\n",
    "            'C_goblet_prog_means','C_goblet_prog_pctexp','SI_goblet_prog_means','SI_goblet_prog_pctexp','C_BEST4_ncells','C_EEC_ncells','C_ISC_ncells','C_TA_ncells','C_earlyCC_ncells','C_goblet_ncells',\n",
    "            'C_goblet_prog_ncells','C_lateCC_ncells','C_tuft_ncells','SI_6-?_ncells','SI_BEST4_ncells','SI_EEC_ncells','SI_FAE_ncells','SI_ISC_ncells','SI_TA2_ncells',\n",
    "            'SI_TA_ncells','SI_earlyAE_ncells','SI_goblet_ncells','SI_goblet_prog_ncells','SI_intermAE_ncells','SI_matureAE_ncells','SI_paneth_ncells','SI_tuft_ncells']\n",
    "concat_list = ['gene_ids','feature_types', 'genome', 'mt']\n",
    "del_list = ['ribo','hb','C_BEST4_means','C_BEST4_pctexp','C_EEC_means','C_EEC_pctexp','C_ISC_means','C_ISC_pctexp','C_TA_means','C_TA_pctexp','C_earlyCC_means',\n",
    "            'C_earlyCC_pctexp','C_goblet_means','C_goblet_pctexp','C_lateCC_means','C_lateCC_pctexp','C_tuft_means','C_tuft_pctexp','SI_BEST4_means','SI_BEST4_pctexp',\n",
    "            'SI_EEC_means','SI_EEC_pctexp','SI_FAE_means','SI_FAE_pctexp','SI_ISC_means','SI_ISC_pctexp','SI_TA2_means','SI_TA2_pctexp','SI_TA_means','SI_TA_pctexp',\n",
    "            'SI_earlyAE_means','SI_earlyAE_pctexp','SI_goblet_means','SI_goblet_pctexp','SI_intermAE_means','SI_intermAE_pctexp','SI_matureAE_means','SI_matureAE_pctexp',\n",
    "            'SI_paneth_means','SI_paneth_pctexp','SI_tuft_means','SI_tuft_pctexp','SI_6-?_means','SI_6-?_pctexp','C_goblet_prog_means','C_goblet_prog_pctexp',\n",
    "            'SI_goblet_prog_means','SI_goblet_prog_pctexp','C_BEST4_ncells','C_EEC_ncells','C_ISC_ncells','C_TA_ncells','C_earlyCC_ncells','C_goblet_ncells',\n",
    "            'C_goblet_prog_ncells','C_lateCC_ncells','C_tuft_ncells','SI_6-?_ncells','SI_BEST4_ncells','SI_EEC_ncells','SI_FAE_ncells','SI_ISC_ncells','SI_TA2_ncells',\n",
    "            'SI_TA_ncells','SI_earlyAE_ncells','SI_goblet_ncells','SI_goblet_prog_ncells','SI_intermAE_ncells','SI_matureAE_ncells','SI_paneth_ncells','SI_tuft_ncells']\n",
    "add_list = ['n_cells','n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts']\n",
    "\n",
    "adata.varm['principle_components'] = adata.varm['PCs']\n",
    "del adata.varm['PCs']\n",
    "\n",
    "for annotation in var_list:\n",
    "    # print(annotation)\n",
    "    if annotation in concat_list:\n",
    "        if annotation == 'gene_ids':\n",
    "            concat_df = adata.var[[a for a in adata.var if annotation in a]].astype('str')\n",
    "            foo = (concat_df['gene_ids-2'].str.upper() == 'NAN')\n",
    "            bar = concat_df['gene_ids-2'].mask(foo,concat_df['gene_ids-1'])\n",
    "            rah = (bar.str.upper() == 'NAN')\n",
    "            dah = bar.mask(rah,concat_df['gene_ids-0'])\n",
    "            adata.var[f'{annotation}'] = dah \n",
    "            del adata.var[f'{annotation}-0'],adata.var[f'{annotation}-1'], adata.var[f'{annotation}-2']\n",
    "        elif annotation == 'mt': \n",
    "            concat_df = adata.var[[a for a in adata.var if annotation in a]].astype('str')\n",
    "            foo = (concat_df['mt-2'].str.upper() == 'NAN')\n",
    "            bar = concat_df['mt-2'].mask(foo,concat_df['mt-1'])\n",
    "            rah = (bar.str.upper() == 'NAN')\n",
    "            dah = bar.mask(rah,concat_df['mt-0'])\n",
    "            adata.var[f'{annotation}'] = dah \n",
    "            del adata.var[f'{annotation}-0'],adata.var[f'{annotation}-1'], adata.var[f'{annotation}-2']\n",
    "        else:\n",
    "            foo = pd.DataFrame({f'{annotation}':adata.var[f'{annotation}-0']}, index = adata.var_names.tolist())\n",
    "            if annotation == 'feature_types': \n",
    "                foo[f'{annotation}'] = 'Gene Expression'\n",
    "                adata.var[f'{annotation}'] = foo\n",
    "                del adata.var[f'feature_types-0'],adata.var[f'{annotation}-1'], adata.var[f'{annotation}-2']\n",
    "            if annotation == 'genome': \n",
    "                foo[f'{annotation}'] = 'GRCh38'\n",
    "                adata.var[f'{annotation}'] = foo\n",
    "                del adata.var[f'{annotation}-0'],adata.var[f'{annotation}-1'], adata.var[f'{annotation}-2']\n",
    "    if annotation in del_list:\n",
    "        if ('means' in annotation) or ('pctexp' in annotation): \n",
    "            del adata.varm[f'{annotation}']\n",
    "        elif ('_ncells' in annotation):\n",
    "            del adata.uns[f'{annotation}']\n",
    "    for i, donor in enumerate(adata.obs['donor'].unique()):\n",
    "        # print(annotation)\n",
    "        if (annotation in del_list) and (f'{annotation}-{i}' in adata.var):\n",
    "            print(annotation)\n",
    "            del adata.var[f'{annotation}-{i}']\n",
    "        if annotation in add_list:\n",
    "            if i == 0:\n",
    "                adata.var[f'Donor_2-{annotation}'] = adata.var[f'{annotation}-{i}']\n",
    "                del adata.var[f'{annotation}-{i}']\n",
    "            if i == 1: \n",
    "                adata.var[f'Donor_1-{annotation}'] = adata.var[f'{annotation}-{i}']\n",
    "                del adata.var[f'{annotation}-{i}']\n",
    "            elif i==2:\n",
    "                adata.var[f'Donor_3-{annotation}'] = adata.var[f'{annotation}-{i}']\n",
    "                del adata.var[f'{annotation}-{i}'] \n",
    "print(adata.uns['neighbors'])\n",
    "\n",
    "d = {'absorptive':'absorptive', 'SI_6-?':'absorptive', 'tuft':'tuft', 'goblet':'goblet', 'BEST4':'BEST4+', 'EEC':'EEC', 'ISC':'ISC', 'TA':'TA',\n",
    "     'nan':'secretory_prog', 'paneth':'paneth', 'FAE':'FAE'}\n",
    "\n",
    "adata.obs.type = adata.obs.type.map(d)\n",
    "\n",
    "print([a for a in adata.obs.type.unique()])\n",
    "\n",
    "pp_adata = mp.load(combined_path, 'preprocessed_adata.h5ad')\n",
    "                   \n",
    "pp_adata = pp_adata[adata.obs_names.tolist()]\n",
    "raw_adata=adata.raw.to_adata()\n",
    "\n",
    "adata.layers['ln_rpdm-normalized'] = raw_adata.copy()\n",
    "adata.layers['raw_rpdm-normalized'] = adata.layers['raw_normalized']\n",
    "del adata.layers['raw_normalized']\n",
    "adata.raw = pp_adata\n",
    "\n",
    "mp.save(adata, combined_path, 'clustered_annotated_adata_k10_lr0.92_v1.6.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import gc\n",
    "data = pd.read_csv(combined_path+'/shared_ENSG_ids_v2.0.csv')\n",
    "adata = mp.load(combined_path, 'clustered_annotated_adata_k10_lr0.92_v1.6.h5ad')\n",
    "\n",
    "raw_adata = adata.raw.to_adata()\n",
    "\n",
    "print('----- Before Modifications ------')\n",
    "\n",
    "print('\\nminimum values')\n",
    "print('raw_adata min: ',raw_adata.X.min())\n",
    "print('raw_min: ',adata.raw.X.min())\n",
    "print('raw_rpdm min: ',adata.layers['raw_rpdm-normalized'].min())\n",
    "print('ln_rpdm min: ',adata.layers['ln_rpdm-normalized'].min())\n",
    "print('scaled min: ',adata.X.min())\n",
    "\n",
    "print('\\nmax values')\n",
    "print('raw_adata max: ',raw_adata.X.max())\n",
    "print('raw_max: ',adata.raw.X.max())\n",
    "print('raw_rpdm max: ',adata.layers['raw_rpdm-normalized'].max())\n",
    "print('ln_rpdm max: ',adata.layers['ln_rpdm-normalized'].max())\n",
    "print('scaled max: ',adata.X.max())\n",
    "\n",
    "del adata.layers['raw_rpdm-normalized']\n",
    "del adata.layers['ln_rpdm-normalized']\n",
    "\n",
    "print(adata)\n",
    "\n",
    "# print(adata[:,[0,1,2,3,11]])\n",
    "\n",
    "for i,geneID in enumerate(data['ENSG ID']):\n",
    "#     print(i+1)\n",
    "#     print(geneID)\n",
    "    \n",
    "    gene_map = {gene:i for i,gene in enumerate(adata.var_names)}\n",
    "    \n",
    "    #generate list of genes to update and most up to date gene symbol\n",
    "    subset = adata[:,adata.var['gene_ids']==geneID].copy()\n",
    "    gc.collect()\n",
    "    genes = subset.var_names.tolist()\n",
    "    # print(genes)\n",
    "    gene_symbol = data.iloc[i]['Current gene symbol']\n",
    "    # print(gene_symbol)\n",
    "    \n",
    "    #Map genes to column position in matrix\n",
    "    gene_positions = [gene_map[genes[0]], gene_map[genes[1]]]\n",
    "    # print(gene_positions)\n",
    "    \n",
    "    #Grab entire layer, update, and save back to adata\n",
    "    # print(type(np.asarray(adata.layers['raw'])))\n",
    "\n",
    "    if i == 0:\n",
    "        layer_ndarray = np.squeeze(np.asarray(adata.layers['raw'].todense()))\n",
    "    else:\n",
    "        layer_ndarray = adata.layers['raw']\n",
    "    \n",
    "    layer_ndarray[:,gene_positions[0]] = layer_ndarray[:,gene_positions].sum(axis=1) #gene_positions[0] - gets overwritten\n",
    "    adata.layers['raw'] = layer_ndarray\n",
    "\n",
    "    #Update gene name\n",
    "    var_names = list(adata.var_names)\n",
    "    var_names[gene_positions[0]] = gene_symbol #gene_positions[0] - name gets replaced\n",
    "    adata.var.index = var_names\n",
    "    raw_adata.var.index = var_names\n",
    "\n",
    "    #Subset to remove second instance\n",
    "    adata = adata[:,[i for i in range(adata.shape[1]) if i != gene_positions[1]]].copy() #gene_positions[1] - gets removed\n",
    "    raw_adata = raw_adata[:,[i for i in range(raw_adata.shape[1]) if i != gene_positions[1]]].copy()\n",
    "    # print(adata.shape)\n",
    "    # print(raw_adata.shape)\n",
    "\n",
    "del adata.raw\n",
    "del adata.layers['raw']\n",
    "\n",
    "adata.var_names_make_unique()\n",
    "\n",
    "adata.layers['raw'] = raw_adata.X\n",
    "\n",
    "results = sc.pp.normalize_total(raw_adata, inplace = False)\n",
    "adata.layers['raw_rpdm-normalized'] = results['X']\n",
    "\n",
    "ln_rpdm_normalized = sc.pp.log1p(adata.layers['raw_rpdm-normalized'], copy = True)\n",
    "adata.layers['ln_rpdm-normalized'] = ln_rpdm_normalized\n",
    "\n",
    "adata.raw = raw_adata\n",
    "\n",
    "print('----- After Modifications ------')\n",
    "\n",
    "print('\\nminimum values')\n",
    "print('raw_adata min: ',raw_adata.X.min())\n",
    "print('raw_min: ',adata.raw.X.min())\n",
    "print('raw_rpdm min: ',adata.layers['raw_rpdm-normalized'].min())\n",
    "print('ln_rpdm min: ',adata.layers['ln_rpdm-normalized'].min())\n",
    "print('scaled min: ',adata.X.min())\n",
    "\n",
    "print('\\nmax values')\n",
    "print('raw_adata max: ',raw_adata.X.max())\n",
    "print('raw_max: ',adata.raw.X.max())\n",
    "print('raw_rpdm max: ',adata.layers['raw_rpdm-normalized'].max())\n",
    "print('ln_rpdm max: ',adata.layers['ln_rpdm-normalized'].max())\n",
    "print('scaled max: ',adata.X.max())\n",
    "\n",
    "mp.save(adata, combined_path, 'clustered_annotated_adata_k10_lr0.92_v1.7.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = mp.load(combined_path, 'clustered_annotated_adata_k10_lr0.92_v1.7.h5ad')\n",
    "\n",
    "print(adata)\n",
    "\n",
    "print('----- After Modifications ------')\n",
    "\n",
    "print('\\nminimum values')\n",
    "print('raw_min: ',adata.raw.X.min())\n",
    "print('raw_rpdm min: ',adata.layers['raw_rpdm-normalized'].min())\n",
    "print('ln_rpdm min: ',adata.layers['ln_rpdm-normalized'].min())\n",
    "print('scaled min: ',adata.X.min())\n",
    "\n",
    "print('\\nmax values')\n",
    "print('raw_max: ',adata.raw.X.max())\n",
    "print('raw_rpdm max: ',adata.layers['raw_rpdm-normalized'].max())\n",
    "print('ln_rpdm max: ',adata.layers['ln_rpdm-normalized'].max())\n",
    "print('scaled max: ',adata.X.max())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = mp.load(combined_path, \"clustered_annotated_adata_k10_lr0.92_v1.6.h5ad\")\n",
    "\n",
    "adata.var['gene_ids'].to_csv(combined_path+'/v1.6_ENSG_values.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "## work in progress updates for standardizing colors for all mappings\n",
    "### Added additional color for lineage=='paneth'\n",
    "type_order = ['absorptive','tuft','goblet','BEST4','EEC','ISC','TA','secretory_prog','paneth','FAE','SI-6?']\n",
    "adata.obs.type = adata.obs.type.astype('category')\n",
    "print(adata.obs.type.unique())\n",
    "adata.obs.type.cat.reorder_categories(type_order, inplace = True)\n",
    "\n",
    "\n",
    "sc.pl.umap(adata, color = ['leiden','lineage','type'], ncols = 1)\n",
    "\n",
    "### Add annotation for Parikh et al 2019 crypt-axis score\n",
    "\n",
    "## Toggle for raw_normalized \n",
    "adata.X = adata.layers['raw_normalized'].copy().todense()\n",
    "print(adata.X.shape)\n",
    "\n",
    "crypt_axis = ['SELENOP','CEACAM7','PLAC8','CEACAM1','TSPAN1','CEACAM5','CEACAM6','IFI27','DHRS9','KRT20',\n",
    "              'RHOC','CD177','PKIB','HPGD','LYPD8']\n",
    "\n",
    "sc.tl.score_genes(adata, crypt_axis, score_name = 'crypt_axis_score_scanpy', use_raw = False)\n",
    "\n",
    "subset = adata[:,crypt_axis].copy()\n",
    "\n",
    "for gene in crypt_axis:\n",
    "    score_set = subset[:,gene].copy()\n",
    "    denom = max(score_set.X)\n",
    "    scale_factor = 1/denom\n",
    "    subset.obs['crypt_axis_score_parikh'] = subset.obs['crypt_axis_score_parikh'] + ((score_set.X.ravel())*scale_factor) #- min(score_set.X.ravel()))* scale_factor)\n",
    "\n",
    "crypt_axis_order = ['C_ISC','C_TA','C_earlyCC','secretory_prog','C_goblet','C_tuft','C_BEST4','EEC','C_lateCC']\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "kernelspec": {
   "display_name": "RNAseq Python",
   "language": "python",
   "name": "rnaseq"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
